File: 0011-container-with-most-water/0011-container-with-most-water.java
class Solution {
    public int maxArea(int[] height) {
        
        // pseudocode
        // we are given an unsorted integer array with numbers representing varying water levels
        // we can use two pointers
        // volume of water = (right pointer - left pointer) * minimum of number at (left pointer , right pointer)
        // initialize pointers
            // left pointer at the left extreme of array
            // right pointer at the right extreme of array
        // initialize a integer variable to contain the maximum volume till current loop
        // check if left pointer < right pointer before entering a loop
        // check the temporary volume and compare with max volume 
            // if max volume < temp volume, assign temp volume to max volume
            // if max volume >= temp volume, move ahead
        // check and compare heights at left and right pointer
            // if height at left pointer <= height at right pointer, left++
            // else right--
        // return max volume
        
        
        //source code
        
        int left = 0;
        int right = height.length - 1;
        int maxVolume = Integer.MIN_VALUE;
        
        while (left < right)
        {
            int tempVolume  = (right - left) * Integer.min(height[left], height[right]);
            maxVolume  = Integer.max(maxVolume, tempVolume);
            if (height[left] <= height[right]) left++;
            else right--;
        }
        
        return maxVolume;
        
    }
}

File: 0011-container-with-most-water/0011-container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        #take two pointers
        #calculate the max water contained in them with the rectangle area
        #the height will be the minimum of the two heights
        #and the width will be the difference between indexes

        l, r = 0, len(height) - 1
        maxWater = 0
        while l < r:
            h = min(height[l], height[r])
            w = r - l
            localwater = h * w
            maxWater = max(maxWater, localwater)
            if height[l] >= height[r]:
                r -= 1
            else:
                l += 1
        return maxWater